### PR TITLE
SI-9121 test case (fixed in new optimizer), SI-9179 test case

### PR DESCRIPTION
**Pull request by lrytz**

### CODE REVIEW

File: src/compiler/scala/tools/nsc/backend/jvm/BCodeHelpers.scala [hunk 1]
```diff
@@ -76,7 +76,7 @@ abstract class BCodeHelpers extends BCodeIdiomatic with BytecodeWriters {
     val origOwner = sym.originalOwner
     // phase travel necessary: after flatten, the name includes the name of outer classes.
     // if some outer name contains $anon, a non-anon class is considered anon.
-    if (delambdafyInline() && sym.rawowner.isAnonymousFunction) {
+    if (delambdafyInline() && exitingPickler(sym.rawowner.isAnonymousFunction)) {
```
Comment (adriaanm) [inline src/compiler/scala/tools/nsc/backend/jvm/BCodeHelpers.scala:79]: Could you add a note on this change to the commit message?

Comment (lrytz) [inline src/compiler/scala/tools/nsc/backend/jvm/BCodeHelpers.scala:79]: done

**Comment by lrytz** [comment]
review by @szeiger

**Comment by lrytz** [comment]
rebased, as there were merge conflicts. ping @szeiger

File: test/junit/scala/tools/nsc/backend/jvm/opt/InlinerTest.scala [hunk 2]
```diff
@@ -1505,4 +1505,25 @@ class InlinerTest extends ClearAfterClass {
...
+        |  implicit class RichFunction1[T1, R](val f: String) extends AnyVal {
+        |    def toRx: Any = ""
+        |  }
+        |}
+        |import p1.Implicits._
+        |class C {
+        |  def t(): Unit = new ScalaObservable("").scMap("")
+        |}
+      """.stripMargin
+    val c :: _ = compile(code)
```
Comment (szeiger) [inline test/junit/scala/tools/nsc/backend/jvm/opt/InlinerTest.scala]: Is a single source unit sufficient to reproduce the issue? I thought it only showed up on separate compilation.

Comment (lrytz) [inline test/junit/scala/tools/nsc/backend/jvm/opt/InlinerTest.scala]: ah, good point!

**Comment by lrytz** [comment]
should be good now

**Comment by szeiger** [comment]
LGTM

**Pull request merged**